pallets/parachain-staking/src/lib.rs

```diff
@@ -837,10 +861,152 @@ pub mod pallet {
+					<AccountsDueUnreservedBalance<T>>::insert(&owner, &account, amount);
+					writes += 1u64;
+					if state.rm_nomination(owner).is_some() {
+						if state.nominations.0.is_empty() {
+							<NominatorState2<T>>::remove(&account);
+						} else {
+							<NominatorState2<T>>::insert(&account, state.clone());
+						}
+						writes += 1u64;
+					}
+				}
+			}
+		}
+		(reads, writes)
+	}
+
+	pub(crate) fn correct_max_nominations_per_collator_upgrade_mistake<T: Config>() -> (u64, u64) {
+		let (mut reads, mut writes) = (0u64, 0u64);
+		for (account, state) in <CollatorState2<T>>::iter() {
+			reads += 1u64;
+			// 1. collect all nominator amounts into single vec and order them
+			let mut all_nominators = state.top_nominators.clone();
+			let mut starting_bottom_nominators = state.bottom_nominators.clone();
+			all_nominators.append(&mut starting_bottom_nominators);
+			// sort all nominators from greatest to least
+			all_nominators.sort_unstable_by(|a, b| b.amount.cmp(&a.amount));
+			let top_n = T::MaxNominatorsPerCollator::get() as usize;
+			// 2. split them into top and bottom using the T::MaxNominatorsPerCollator
+			let top_nominators: Vec<Bond<T::AccountId, BalanceOf<T>>> =
+				all_nominators.clone().into_iter().take(top_n).collect();
+			let bottom_nominators = if all_nominators.len() > top_n {
+				let rest = all_nominators.len() - top_n;
+				let bottom: Vec<Bond<T::AccountId, BalanceOf<T>>> = all_nominators
+					.clone()
+					.into_iter()
+					.rev()
+					.take(rest)
+					.collect();
+				bottom
+			} else {
+				// empty, all nominations are in top
+				Vec::new()
+			};
+			let (mut total_counted, mut total_backing): (BalanceOf<T>, BalanceOf<T>) =
+				(0u32.into(), 0u32.into());
+			for Bond { amount, .. } in &top_nominators {
+				total_counted += *amount;
+				total_backing += *amount;
+			}
+			for Bond { amount, .. } in &bottom_nominators {
+				total_backing += *amount;
+			}
+			// update candidate pool with new total counted if it changed
+			if state.total_counted != total_counted && state.is_active() {
+				reads += 1u64;
+				writes += 1u64;
+				<Pallet<T>>::update_active(account.clone(), total_counted);
+			}
+			<CollatorState2<T>>::insert(
+				account,
+				Collator2 {
+					top_nominators,
+					bottom_nominators,
+					total_counted,
+					total_backing,
+					..state
+				},
+			);
+			writes += 1u64;
+		}
+		(reads, writes)
 	}
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let (mut reads, mut writes) =
+				correct_bond_less_removes_bottom_nomination_inconsistencies::<T>();
+			let (reads_1, writes_1) = correct_max_nominations_per_collator_upgrade_mistake::<T>();
+			reads += reads_1;
+			writes += writes_1;
+			let wt = <T as frame_system::Config>::DbWeight::get();
+			wt.reads(reads) + wt.writes(writes) + 10_000_000_000 // 2% of the max block weight margin
```

Comment:
on_runtime_upgrade should be replayable. So you need to check if the migration is necessary or not. And if the is already played, return only the weight needed to check if the migration is already done.
The best way to do that is to add a StorageValue that store the pallet version, like the balance pallet do.

Reply:
For instance, you can do that:

```rust
// A value placed in storage that represents the current version of the Balances storage.
// This value is used by the `on_runtime_upgrade` logic to determine whether we run
// storage migration logic.
#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, MaxEncodedLen)]
enum StorageVersions {
	V1,
	V2,
}

/// Storage version of the pallet.
#[pallet::storage]
pub(super) type StorageVersion<T: Config> = StorageValue<_, Releases, ValueQuery>;
```

Then impl Default for `StorageVersions` enum tu return `V1` when this storage entry not exist.

Reply:
Good point, we need something here to prevent it from running twice. In the past I've used a `bool` in storage to track if the migration has executed yet.

We've discussed adding a `StorageVersion` storage item like this in the past, but the discussion criticized the lack of flexibility relative to semver so I gave up on it. We can rediscuss at the meeting today.

Reply:
> but the discussion criticized the lack of flexibility relative to semver so I gave up on it.

There is no obligation to use semver for StorageVersion. It can be a simple integer that is incremented at each migration.

Reply:
> We can rediscuss at the meeting today.

Alan wants us to merge this quickly, I think we can't wait for our meeting, I think it's better that you choose now a simple solution to manage the replay of on-runtime-upgrade

Reply:
@librelois I added a simple bool in storage to track if the migration has run yet. Let me know if it is sufficient.

Reply:
> wt.reads(reads) + wt.writes(writes) + 10_000_000_000 //2% of the max block weight margin 10_000_000_000

Why keep a base weight of 10_000_000_000 in cases where nothing is done? Put all this in the if ;)

Reply:
Ok now it's good